Paste product combination tokens and highlight invalid tokens

MarkALeonard opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow product combination tokens to be pasted. Pasted comma separated text will be validated and tokens added. Invalid tokens will be coloured yellow. Duplicate tokens will not be added.
| Type?         |  improvement 
| Category?     |  BO
| BC breaks?    |  no
| Deprecations? |  no
| Fixed ticket? | Fixes #{issue number here}.
| How to test?  | **On Current Product Page:**  Copy and paste 'Color : Red, Color : Blue, Color : Red, Color : Green, Color : Invalid' into the product combination tokens field. All tokens will be created including duplicates but no attributes will be selected and pressing the generate button produces a single product combination. **On Product Page with this Improvement:** Copy and paste 'Color : Red, Color : Blue, Color : Red, Color : Green, Color : Invalid' into the product combination tokens field. Tokens will be created, duplicate tokens will be removed, attributes will be selected and invalid tokens highlighted in yellow. Pressing the generate button will create the correct number of product combinations.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/18095)
<!-- Reviewable:end -->

MarkALeonard committed:
Allow copy and paste combination tokens

MarkALeonard committed:
Stop duplicate token creation

MarkALeonard committed:
Fixed invalid token filtering

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,19 +929,51 @@ var form = (function() {
+              }
+            });
+          } else {
+            var attr = $('.js-attribute-checkbox[data-value="' + e.attrs.value + '"]');
```
```suggestion
            var attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,19 +929,51 @@ var form = (function() {
+          } else {
+            var attr = $('.js-attribute-checkbox[data-value="' + e.attrs.value + '"]');
+
+            if (attr != undefined) {
```
```suggestion
            if (attr) {
```

Wouldn't be just enough ? Else prefer using strict operators anyway

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,19 +929,51 @@ var form = (function() {
+
+          if(e.attrs.data){
+            if(filter([e.attrs]).length == 0){
+              $('#form_step3_attributes-tokenfield').val(function(i, value){
```
I've a doubt, this file is processed with webpack right? If yes then prefer using arrow functions
```suggestion
              $('#form_step3_attributes-tokenfield').val((i, value) => {
```

NeOMakinG commented:
Thanks for your contribution @MarkALeonard, I left some comments mainly aiming good practices :)

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard commented:
Thanks @NeOMakinG for your feedback, I have made the changes accordingly.

NeOMakinG commented:
> Thanks @NeOMakinG for your feedback, I have made the changes accordingly.

Did you test it after changes ?  Hope the `if (attr)` is fine in most case, thanks for you work.

MarkALeonard commented:
Yes, tested and works as expected. Thank you.

MarkALeonard committed:
Fix errors on token removal.
Imorove multiple engine search results.

MarkALeonard commented:
On implementing on a live site I noticed the Bloodhound engine returns multiple results for 'Size : S' but not 'Size : M'. Have changed `result.length == 1` to `result.length >= 1` and first result in array is used. Also fixed issue from original code that reported an error when removing invalid tokens. As the new code has the 'invalid' class I have added a check for that.

Progi1984 commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
           return false;
         }
+
+        if (!e.attrs.data){
```
```suggestion
        if (!e.attrs.data)
          if (e.handleObj.origType !== 'tokenfield:createtoken') {
            return false;
          }
```

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: Progi1984 <progi1984@gmail.com>

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,58 @@ var form = (function() {
+          }
+
+          if(e.attrs.data){
+            if(filter([e.attrs]).length == 0){
```
```suggestion
            if(filter([e.attrs]).length === 0){
```

In that case that's unnecessary but as we want to keep consistance in our codebase, can you prefer using strict operators ?

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,58 @@ var form = (function() {
+            return false;
+          }
+          var orgLabel = e.attrs.label;
+          if (e.attrs.label == e.attrs.value) {
```
```suggestion
          if (e.attrs.label === e.attrs.value) {
```

NeOMakinG commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,58 @@ var form = (function() {
       $('#form_step3_attributes').on('tokenfield:removedtoken', function(e) {
-        $('#attribute-generator-' + e.attrs.value).remove();
+        if(!$(e.relatedTarget).hasClass('invalid')){
+          $('#attribute-generator-' + e.attrs.value).remove();
```
```suggestion
          $(`#attribute-generator-${e.attrs.value}`).remove();
```

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: SZCZUPAK Valentin <valentin.szczupak@prestashop.com>

MarkALeonard committed:
Removed unrequired closing brace

NeOMakinG reviewed (dismissed):
Thanks!

PierreRambaud commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
+        if (!e.attrs.data)
+          if (e.handleObj.origType !== 'tokenfield:createtoken') {
+            return false;
+          }
```
```suggestion
        if (!e.attrs.data && e.handleObj.origType !== 'tokenfield:createtoken') {
          return false;
        }
```

PierreRambaud replied:
Could you explain why you split the if statement?

MarkALeonard replied:
The statement was split as the next section of the code only needs to run if the data attributes have not been set. The revision was suggested earlier in this converstion. However I have noticed that with the change including the dropping of braces it actually doesn't work as intended. I will restructure it and push a new version which will make the decision clearer.

PierreRambaud commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
+            return false;
+          }
+          
+        var orgLabel = e.attrs.label;
```
```suggestion
        const orgLabel = e.attrs.label;
```

PierreRambaud commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
+            }
+          });
+        } else {
+          var attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```
```suggestion
          const attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```

PierreRambaud commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
+            });
+            return false;
+          };
+        }
```
```suggestion
      if(e.attrs.data && filter([e.attrs]).length === 0){
          $('#form_step3_attributes-tokenfield').val((i, value) => {
            return value.replace(orgLabel,"");
          });
          return false;
        }
```

PierreRambaud commented on admin-dev/themes/default/js/bundle/product/form.js:
```diff
@@ -929,25 +929,59 @@ var form = (function() {
       /** On event "tokenfield:removedtoken" : remove stored attributes input when remove token */
       $('#form_step3_attributes').on('tokenfield:removedtoken', function(e) {
-        $('#attribute-generator-' + e.attrs.value).remove();
+        if(!$(e.relatedTarget).hasClass('invalid')){
```
```suggestion
        if (!$(e.relatedTarget).hasClass('invalid')) {
```

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

MarkALeonard committed:
Update admin-dev/themes/default/js/bundle/product/form.js

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

MarkALeonard committed:
Add braces to if to enclose code.

khouloudbelguith commented:
Hi @MarkALeonard,

Can you please resolve conflicts.

Thanks!

MarkALeonard committed:
Merge branch 'develop' into paste-combinations

SD1982 commented:
LGTM Thanks @MarkALeonard

Progi1984 commented:
Thanks @MarkALeonard
